### DISCUSSION TITLE
Slow index request because fields are loaded despite defined fieldsForIndex

### DISCUSSION THREAD

**Discussion by m-lotze** [laravel/nova-issues#4507] [General]
- Laravel Version: 9.19.0
- Nova Version: 4.11.0

### Description:
When a resource defines `fieldsForIndex` the `fields` method is still executed for each resource on an index request. When you have a lot of fields and e.g. load relations and so on, this slows down the index request a lot. In our case changing the `fields` method like this improves the loading speed quite a bit:

```
public function fields(NovaRequest $request): array
{
    if ($request->isResourceIndexRequest()) {
        return [];
    }

   ...
}
```

I assume this is also the case for other methods like `fieldsForDetail`. I think Nova should only load `fields` if there are no other methods defined for the specific request.

**Comment by crynobone** [comment]
We uses `fields` to verify multiple logic but that's shouldn't have much affect. My guess you have some heavy processing code running under `fields()` which causes the performance issues.

**Comment by juse-less** [comment]
Depending on where you load the relationships, I recommend implementing `Resource::indexQuery()`, and/or `Resource::detailQuery()`.

I couldn't find this directly referenced in the documentation right now, but you can see some examples for both pre-defined query methods, as well as dynamic ones at [documentaion/resources/authorization](https://nova.laravel.com/docs/4.0/resources/authorization.html).
Like [indexQuery](https://nova.laravel.com/docs/4.0/resources/authorization.html#index-filtering), and [relatableQuery](https://nova.laravel.com/docs/4.0/resources/authorization.html#relatable-filtering) (which can be dynamically implemented).

Also keep in mind - if you'll _always_ load relationships, regardless of page 'type', you can define the `$with` property, just like in Models.

```php

// Keep in mind - only BelongsTo (and equivalent) relationships are presented on indexes.
// If you need to present a summary of a HasMany relationship, or so, you can load that as well.
public static function indexQuery(NovaRequest $request, $query)
{
    return parent::indexQuery($request, $query->with([
        'user',
        'someBelongsTo',
        'otherRelation',
        // ...
    ]));
}

// All relationships can be presented.
public static function detailQuery(NovaRequest $request, $query)
{
    return parent::detailQuery($request, $query->with([
        'user',
        'someHasMany',
        'otherRelation',
        // ...
    ]));
}
```
